File: bindings/dotnet/src/Pamoja.Core/Power.cs
using Pamoja.Core.Interop;

namespace Pamoja.Core;

/// <summary>What a node should be doing at the current state of charge.</summary>
public enum PowerMode
{
    /// <summary>Full duty, because the charge is healthy.</summary>
    Active = 0,

    /// <summary>Reduced duty, to conserve charge.</summary>
    Saver = 1,

    /// <summary>Minimum duty, to stay alive as long as possible.</summary>
    Critical = 2,
}

/// <summary>The split between the time a node works and the time it sleeps.</summary>
/// <param name="ActiveUs">How long the node stays awake each period, in microseconds.</param>
/// <param name="SleepUs">How long it sleeps each period, in microseconds.</param>
public readonly record struct DutyCycle(ulong ActiveUs, ulong SleepUs)
{
    /// <summary>Creates a duty cycle that spends a fraction of a period awake.</summary>
    /// <param name="periodUs">The whole period, in microseconds.</param>
    /// <param name="fraction">The share spent awake, clamped to 0 through 1.</param>
    /// <returns>The duty cycle.</returns>
    public static DutyCycle FromFraction(ulong periodUs, float fraction)
    {
        PamojaDutyCycle duty = NativeMethods.pamoja_duty_cycle_from_fraction(periodUs, fraction);
        return new DutyCycle(duty.ActiveUs, duty.SleepUs);
    }

    /// <summary>Gets the whole period, awake plus asleep, in microseconds.</summary>
    public ulong PeriodUs => NativeMethods.pamoja_duty_cycle_period_us(Native);

    /// <summary>Gets the share of the period spent awake, from 0 through 1.</summary>
    public float Fraction => NativeMethods.pamoja_duty_cycle_fraction(Native);

    /// <summary>Gets the blittable form the C ABI takes.</summary>
    private PamojaDutyCycle Native => new() { ActiveUs = ActiveUs, SleepUs = SleepUs };
}

/// <summary>The work intervals a node uses in each mode, and where they change.</summary>
/// <remarks>
/// A node on a battery and a panel has to decide how often to do anything at all.
/// A plan stretches the interval between work as the charge falls, so a node that
/// would otherwise go dark in a cloudy week keeps reporting, less often.
/// </remarks>
/// <param name="ActiveUs">The interval at a healthy charge, in microseconds.</param>
/// <param name="SaverUs">The longer interval used to conserve, in microseconds.</param>
/// <param name="CriticalUs">The longest interval, in microseconds.</param>
/// <param name="SaverBelow">Enter <see cref="PowerMode.Saver"/> below this charge.</param>
/// <param name="CriticalBelow">Enter <see cref="PowerMode.Critical"/> below this charge.</param>
public readonly record struct PowerPlan(
    ulong ActiveUs,
    ulong SaverUs,
    ulong CriticalUs,
    float SaverBelow,
    float CriticalBelow)
{
    /// <summary>Creates a plan from its three work intervals, with the defaults.</summary>
    /// <remarks>
    /// The defaults enter <see cref="PowerMode.Saver"/> below 50% charge and
    /// <see cref="PowerMode.Critical"/> below 20%.
    /// </remarks>
    /// <param name="activeUs">The interval at a healthy charge, in microseconds.</param>
    /// <param name="saverUs">The longer interval used to conserve, in microseconds.</param>
    /// <param name="criticalUs">The longest interval, in microseconds.</param>
    /// <returns>The power plan.</returns>
    public static PowerPlan Create(ulong activeUs, ulong saverUs, ulong criticalUs)
    {
        PamojaPowerPlan plan =
            NativeMethods.pamoja_power_plan_new(activeUs, saverUs, criticalUs);
        return new PowerPlan(
            plan.ActiveUs,
            plan.SaverUs,
            plan.CriticalUs,
            plan.SaverBelow,
            plan.CriticalBelow);
    }

    /// <summary>Returns a plan with the state-of-charge thresholds moved.</summary>
    /// <param name="saverBelow">Enter saver mode below this charge.</param>
    /// <param name="criticalBelow">Enter critical mode below this charge.</param>
    /// <returns>The adjusted plan.</returns>
    public PowerPlan WithThresholds(float saverBelow, float criticalBelow)
    {
        PamojaPowerPlan plan = NativeMethods.pamoja_power_plan_with_thresholds(
            Native, saverBelow, criticalBelow);
        return new PowerPlan(
            plan.ActiveUs,
            plan.SaverUs,
            plan.CriticalUs,
            plan.SaverBelow,
            plan.CriticalBelow);
    }

    /// <summary>Returns the mode this plan calls for at a state of charge.</summary>
    /// <param name="soc">The battery state of charge, from 0 through 1.</param>
    /// <returns>The mode the node should run in.</returns>
    public PowerMode Mode(float soc) =>
        (PowerMode)NativeMethods.pamoja_power_plan_mode(Native, soc);

    /// <summary>Returns the mode, eased one step toward full duty while charging.</summary>
    /// <param name="soc">The battery state of charge, from 0 through 1.</param>
    /// <param name="charging">Whether the node is taking charge.</param>
    /// <returns>The mode the node should run in.</returns>
    public PowerMode ModeWhileCharging(float soc, bool charging) =>
        (PowerMode)NativeMethods.pamoja_power_plan_mode_while_charging(
            Native, soc, charging ? (byte)1 : (byte)0);

    /// <summary>Returns the work interval for a mode, in microseconds.</summary>
    /// <param name="mode">The mode to look up.</param>
    /// <returns>The interval in microseconds.</returns>
    public ulong IntervalForUs(PowerMode mode) =>
        NativeMethods.pamoja_power_plan_interval_for_us(Native, (PamojaPowerMode)mode);

    /// <summary>Returns the work interval at a state of charge, in microseconds.</summary>
    /// <param name="soc">The battery state of charge, from 0 through 1.</param>
    /// <returns>The interval in microseconds.</returns>
    public ulong IntervalUs(float soc) =>
        NativeMethods.pamoja_power_plan_interval_us(Native, soc);

    /// <summary>Gets the blittable form the C ABI takes.</summary>
    private PamojaPowerPlan Native => new()
    {
        ActiveUs = ActiveUs,
        SaverUs = SaverUs,
        CriticalUs = CriticalUs,
        SaverBelow = SaverBelow,
        CriticalBelow = CriticalBelow,
    };
}

File: examples/tests/guides/power.rs
//! The power-budget guide example; see docs/guides/power.md.

/// A solar node deciding how often to wake as its battery falls, and what that cadence
/// costs it in average draw.
#[test]
fn a_falling_charge_stretches_the_work_interval() {
    // ANCHOR: example
    use core::time::Duration;

    use pamoja_power::{DutyCycle, PowerMode, PowerPlan};

    // A solar node samples every minute while the charge is healthy, stretches to ten
    // minutes to conserve, and to an hour once the battery is nearly flat.
    let plan = PowerPlan::new(
        Duration::from_secs(60),
        Duration::from_secs(600),
        Duration::from_secs(3600),
    );

    // The default thresholds enter saver mode below 50% charge and critical below 20%.
    for charge in [0.80, 0.35, 0.12] {
        let mode = plan.mode(charge);
        let every = plan.interval(charge).as_secs();
        let percent = charge * 100.0;
        println!("at {percent:.0}% charge: {mode:?}, sampling every {every}s");
    }

    // A panel that is delivering buys back one mode, so the same flat battery keeps
    // reporting on the ten-minute saver cadence while the sun is on it.
    let charging = plan.mode_while_charging(0.12, true);
    println!("the same flat battery, while charging: {charging:?}");

    // The work is the same two seconds whichever mode the node is in; stretching the cycle
    // is what saves the energy. The duty fraction is the proxy for average draw, so the
    // hourly cadence costs a sixtieth of what the one-minute cadence does.
    let awake = Duration::from_secs(2);
    let healthy = DutyCycle::new(awake, plan.interval(0.80) - awake);
    let flat = DutyCycle::new(awake, plan.interval(0.12) - awake);
    let (healthy_duty, flat_duty) = (healthy.fraction() * 100.0, flat.fraction() * 100.0);
    println!("awake {healthy_duty:.2}% of the time when healthy");
    println!("awake {flat_duty:.3}% of the time when flat");

    // Stating the budget as a fraction instead gives the awake time directly.
    let quarter = DutyCycle::from_fraction(Duration::from_secs(1), 0.25);
    println!("a quarter-duty second is {:?} awake", quarter.active());
    // ANCHOR_END: example

    assert_eq!(plan.mode(0.80), PowerMode::Active);
    assert_eq!(plan.interval(0.80), Duration::from_secs(60));
    assert_eq!(plan.mode(0.35), PowerMode::Saver);
    assert_eq!(plan.mode(0.12), PowerMode::Critical);
    assert_eq!(plan.interval(0.12), Duration::from_secs(3600));
    assert_eq!(charging, PowerMode::Saver);
    assert!((healthy.fraction() - 2.0 / 60.0).abs() < 1e-6);
    assert!((flat.fraction() - 2.0 / 3600.0).abs() < 1e-6);
    assert_eq!(quarter.active(), Duration::from_millis(250));
    assert_eq!(quarter.sleep(), Duration::from_millis(750));
}
